File: examples/go/production_workflow.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/yourdudeken/mpesa-sdk/client"
	"github.com/yourdudeken/mpesa-sdk/types"
)

func main() {
	mpesa := client.NewClient(types.MpesaConfig{
		ConsumerKey:        os.Getenv("MPESA_CONSUMER_KEY"),
		ConsumerSecret:     os.Getenv("MPESA_CONSUMER_SECRET"),
		Environment:        types.Production,
		Passkey:            os.Getenv("MPESA_PASSKEY"),
		InitiatorName:      os.Getenv("MPESA_INITIATOR_NAME"),
		SecurityCredential: os.Getenv("MPESA_SECURITY_CREDENTIAL"),
		Timeout:            30 * time.Second,
	})

	ctx := context.Background()

	// 1. Initiate STK Push
	resp, err := mpesa.STKPush(ctx, types.STKPushRequest{
		BusinessShortCode: 174379,
		TransactionType:   types.CustomerPayBillOnline,
		Amount:            5000,
		PartyA:            254722000000,
		PartyB:            174379,
		PhoneNumber:       254722000000,
		CallBackURL:       "https://api.yourdomain.com/mpesa/callback",
		AccountReference:  "ORDER-12345",
		TransactionDesc:   "Order payment",
	})
	if err != nil {
		log.Fatalf("STK Push failed: %v", err)
	}
	fmt.Printf("1. STK Push sent: %s\n", resp.CheckoutRequestID)

	// 2. Query STK status after delay
	time.Sleep(5 * time.Second)
	status, err := mpesa.STKQuery(ctx, types.STKQueryRequest{
		BusinessShortCode: "174379",
		CheckoutRequestID: resp.CheckoutRequestID,
	})
	if err != nil {
		log.Printf("2. Query failed: %v", err)
	} else {
		fmt.Printf("2. Payment status: %s\n", status.ResultDesc)
	}

	// 3. Check account balance
	balance, err := mpesa.AccountBalance(ctx, types.AccountBalanceRequest{
		Initiator:          os.Getenv("MPESA_INITIATOR_NAME"),
		SecurityCredential: os.Getenv("MPESA_SECURITY_CREDENTIAL"),
		CommandID:          "AccountBalance",
		PartyA:             174379,
		IdentifierType:     4,
		Remarks:            "Daily reconciliation",
		QueueTimeOutURL:    "https://api.yourdomain.com/mpesa/queue",
		ResultURL:          "https://api.yourdomain.com/mpesa/balance-result",
	})
	if err != nil {
		log.Printf("3. Balance query failed: %v", err)
	} else {
		fmt.Printf("3. Balance query sent: %s\n", balance.OriginatorConversationID)
	}
}

File: examples/typescript/production-workflow.ts
import { Mpesa } from "mpesa-sdk";

const RAW_SHORTCODE = process.env.MPESA_SHORTCODE;
if (!RAW_SHORTCODE) throw new Error("MPESA_SHORTCODE is required");
const SHORTCODE = parseInt(RAW_SHORTCODE, 10);
if (isNaN(SHORTCODE) || SHORTCODE <= 0) {
  throw new Error(`Invalid MPESA_SHORTCODE: ${RAW_SHORTCODE}`);
}

const mpesa = new Mpesa({
  consumerKey: process.env.MPESA_CONSUMER_KEY!,
  consumerSecret: process.env.MPESA_CONSUMER_SECRET!,
  environment: "production",
  passkey: process.env.MPESA_PASSKEY!,
  timeout: 30000,
  retryConfig: {
    maxRetries: 3,
    baseDelayMs: 2000,
    maxDelayMs: 60000,
  },
  logging: {
    onRequest: (req) => console.log("→", req.method, req.url),
    onError: (err) => console.error("✗", err),
  },
});

export async function productionWorkflow() {
  const stkResponse = await mpesa.stkPush.initiate({
    BusinessShortCode: SHORTCODE,
    TransactionType: "CustomerPayBillOnline",
    Amount: 5000,
    PartyA: 254722000000,
    PartyB: SHORTCODE,
    PhoneNumber: 254722000000,
    CallBackURL: "https://api.yourdomain.com/mpesa/callback",
    AccountReference: "ORDER-12345",
    TransactionDesc: "Order payment",
    Password: "",
    Timestamp: "",
  });
  console.log("STK Push sent:", stkResponse.CheckoutRequestID);

  await new Promise((r) => setTimeout(r, 5000));
  const status = await mpesa.stkPush.query({
    BusinessShortCode: String(SHORTCODE),
    CheckoutRequestID: stkResponse.CheckoutRequestID,
    Password: "",
    Timestamp: "",
  });
  console.log("Payment status:", status.ResultDesc);

  const balance = await mpesa.accountBalance.query({
    Initiator: process.env.MPESA_INITIATOR_NAME!,
    SecurityCredential: process.env.MPESA_SECURITY_CREDENTIAL!,
    CommandID: "AccountBalance",
    PartyA: SHORTCODE,
    IdentifierType: 4,
    Remarks: "Reconcile balance",
    QueueTimeOutURL: "https://api.yourdomain.com/mpesa/queue",
    ResultURL: "https://api.yourdomain.com/mpesa/balance-result",
  });
  console.log("Balance query sent:", balance.OriginatorConversationID);
}

if (process.env.RUN_PRODUCTION_WORKFLOW === "1") {
  productionWorkflow().catch((err) => console.error("Workflow failed:", err));
}
